Sources/GRPCCore/Metadata.swift

```diff
@@ -14,5 +14,267 @@
+    self.lockedElements.withLockedValue { elements in
+      elements.removeAll { metadataKeyValue in
+        metadataKeyValue.key == key
+      }
+      elements.append(.init(key: key, value: .string(stringValue)))
+    }
+  }
+
+  /// Adds a key-value pair to the collection, where the value is `Data`.
+  /// If there are pairs already associated to the given key, they will all be removed first, and the new pair
+  /// will be added. If no pairs are present with the given key, a new one will be added.
+  /// - Parameters:
+  ///   - key: The key to be associated with the given value. Must end in `-bin`.
+  ///   - binaryValue: The `Data` to be associated with the given key.
+  public mutating func replaceOrAdd(key: String, binaryValue: Data) {
+    self.lockedElements.withLockedValue { elements in
+      elements.removeAll { metadataKeyValue in
+        metadataKeyValue.key == key
+      }
+      elements.append(.init(key: key, value: .binary(binaryValue)))
+    }
+  }
+
+  /// Adds a key-value pair to the collection.
+  /// If there are pairs already associated to the given key, they will all be removed first, and the new pair
+  /// will be added. If no pairs are present with the given key, a new one will be added.
+  /// - Parameters:
+  ///   - key: The key to be associated with the given value. If value is binary, it must end in `-bin`.
+  ///   - value: The ``MetadataValue`` to be associated with the given key.
+  public mutating func replaceOrAdd(key: String, value: MetadataValue) {
+    self.lockedElements.withLockedValue { elements in
+      elements.removeAll { metadataKeyValue in
+        metadataKeyValue.key == key
+      }
+      elements.append(.init(key: key, value: value))
+    }
+  }
+}
+
+extension Metadata: RandomAccessCollection {
+  public typealias Element = MetadataKeyValue
+
+  public struct Index: Comparable, Sendable {
+    @usableFromInline
+    let _base: Array<Element>.Index
+
+    @inlinable
+    init(_base: Array<Element>.Index) {
+      self._base = _base
+    }
+
+    @inlinable
+    public static func < (lhs: Index, rhs: Index) -> Bool {
+      return lhs._base < rhs._base
+    }
+  }
+
+  public var startIndex: Index {
+    return .init(_base: self.elements.startIndex)
+  }
+
+  public var endIndex: Index {
+    return .init(_base: self.elements.endIndex)
+  }
+
+  public func index(before i: Index) -> Index {
+    return .init(_base: self.elements.index(before: i._base))
+  }
+
+  public func index(after i: Index) -> Index {
+    return .init(_base: self.elements.index(after: i._base))
+  }
+
+  public subscript(position: Index) -> Element {
+    self.elements[position._base]
+  }
+}
+
+extension Metadata {
+
+  /// An iterator for all string values associated with a given key.
+  /// This iterator will only return values originally stored as strings for a given key.
+  public struct StringValuesIterator: IteratorProtocol {
+    private var metadataIterator: Metadata.Iterator
```

Comment:
To complement the string and binary sequences we should also have a sequence which provides `MetadataValue` values for a given key. We can use that iterator as the base here.